runtime/plaid/src/functions/storage.rs

```diff
@@ -229,6 +343,145 @@ pub fn get(
+    key_buf_len: u32,
+    data_buffer: WasmPtr<u8>,
+    data_buffer_len: u32,
+) -> i32 {
+    let store = env.as_store_ref();
+    let env_data = env.data();
+
+    let storage = if let Some(storage) = &env_data.storage {
+        storage
+    } else {
+        return FunctionErrors::ApiNotConfigured as i32;
+    };
+
+    let memory_view = match get_memory(&env, &store) {
+        Ok(memory_view) => memory_view,
+        Err(e) => {
+            error!(
+                "{}: Memory error in storage_get: {:?}",
+                env_data.module.name, e
+            );
+            return FunctionErrors::CouldNotGetAdequateMemory as i32;
+        }
+    };
+
+    let key = match safely_get_string(&memory_view, key_buf, key_buf_len) {
+        Ok(s) => s,
+        Err(e) => {
+            error!(
+                "{}: Key error in storage_get: {:?}",
+                env_data.module.name, e
+            );
+            return FunctionErrors::ParametersNotUtf8 as i32;
+        }
+    };
+
+    get_common(
+        env_data,
+        storage,
+        &env_data.module.name,
+        &key,
+        memory_view,
+        data_buffer,
+        data_buffer_len,
+    )
+}
+
+/// Get data from a shared namespace in the storage system, if one is configured
+pub fn get_shared(
+    env: FunctionEnvMut<Env>,
+    namespace_buf: WasmPtr<u8>,
+    namespace_buf_len: u32,
+    key_buf: WasmPtr<u8>,
+    key_buf_len: u32,
+    data_buffer: WasmPtr<u8>,
+    data_buffer_len: u32,
+) -> i32 {
+    let store = env.as_store_ref();
+    let env_data = env.data();
+
+    let storage = if let Some(storage) = &env_data.storage {
+        storage
+    } else {
+        return FunctionErrors::ApiNotConfigured as i32;
+    };
+
+    let memory_view = match get_memory(&env, &store) {
+        Ok(memory_view) => memory_view,
+        Err(e) => {
+            error!(
+                "{}: Memory error in storage_get_shared: {:?}",
+                env_data.module.name, e
+            );
+            return FunctionErrors::CouldNotGetAdequateMemory as i32;
+        }
+    };
+
+    let namespace = match safely_get_string(&memory_view, namespace_buf, namespace_buf_len) {
+        Ok(s) => s,
+        Err(e) => {
+            error!(
+                "{}: Namespace error in storage_get_shared: {:?}",
+                env_data.module.name, e
+            );
+            return FunctionErrors::ParametersNotUtf8 as i32;
+        }
+    };
+
+    let key = match safely_get_string(&memory_view, key_buf, key_buf_len) {
+        Ok(s) => s,
+        Err(e) => {
+            error!(
+                "{}: Key error in storage_get_shared: {:?}",
+                env_data.module.name, e
+            );
+            return FunctionErrors::ParametersNotUtf8 as i32;
+        }
+    };
+
+    // Check if the rule can read from the shared DB.
+    // This is the case if the rule is in either the "r" or "rw" section of the config.
+    let allowed = match &storage.shared_dbs {
```

Comment:
I would split this check up because we can move if this is `None` to the top of function, before any memory operations.